### PR TITLE
Add Split.IO support to the czgenepi backend

### PR DESCRIPTION
**Pull request by jgadling**
### Summary:
- **What:** `<brief description>`
- **Ticket:** [sc166024](https://app.shortcut.com/genepi/story/166024)
- **Env:** https://splitbe-frontend.dev.czgenepi.org/

### Notes:
This sets up split to integrate nicely into fastapi endpoints as a dependency like so:
```
splitio: SplitClient = Depends(get_splitio)
```

The value for a feature flag can then be fetched with:
```
flag_value = splitio.get_flag(user, "flag_name_here")
```

If the value of the split api key is `localhost` (which is the default at the moment since I haven't updated our staging/prod secrets), split will read feature flag configuration from a file at `src/backend/etc/split.yml`. The yaml format is [documented here](https://help.split.io/hc/en-us/articles/360020359652-Python-SDK#localhost-mode)

### Checklist:
- [x] I merged latest `<base branch>`
- [x] I manually verified the change
- [x] I added labels to my PR
- [ ] I tested in multiple browsers
- [ ] I added relevant unit tests
- [ ] I have notified others of changes they need to make locally (migrations, jobs, package updates, etc)

### CODE REVIEW

File: src/backend/Dockerfile [hunk 1]
```diff
@@ -24,10 +39,22 @@ WORKDIR /usr/src/app
 
 COPY pyproject.toml poetry.lock environment.yaml ./
 COPY third-party ./third-party
-RUN /opt/poetry/bin/poetry config virtualenvs.create false && \
-    /opt/poetry/bin/poetry install
+RUN /opt/poetry/bin/poetry config virtualenvs.create false
+# Copy poetry deps from the build image
+COPY --from=build /usr/local/lib/python3.9/site-packages /usr/local/lib/python3.9/site-packages
```
Comment (jgadling) [inline src/backend/Dockerfile:44]: This is just an attempt to trim down our image a little bit, borrowing from some of the tricks used in the nexstrain/base image build.

File: src/backend/aspen/api/auth.py [hunk 2]
```diff
@@ -70,12 +70,12 @@ async def get_auth_user(
     # Redirect to Login page
     if not auth0_user_id:
         # TODO - redirect to login.
-        raise Exception("No Userid Found!")
+        raise ex.UnauthenticatedException("Login failure")
```
Comment (jgadling) [inline src/backend/aspen/api/auth.py:73]: Get more specific about what kinds of errors we're raising

File: src/backend/aspen/api/error/http_exceptions.py [hunk 3]
```diff
@@ -27,10 +27,14 @@ async def exception_handler(request: Request, exc: AspenException) -> JSONRespon
...
 
-class UnauthorizedException(AspenException):
+class UnauthenticatedException(AspenException):
     status_code = 401
 
 
+class UnauthorizedException(AspenException):
+    status_code = 403
+
+
```
Comment (jgadling) [inline src/backend/aspen/api/error/http_exceptions.py:37]: Two different kinds of auth errors

File: src/backend/aspen/api/main.py [hunk 4]
```diff
@@ -51,6 +52,12 @@ def get_app() -> FastAPI:
     # Add a global settings object to the app that we can use as a dependency
     _app.state.aspen_settings = settings
 
+    # Set up Split.io feature flagging
+    splitio = SplitClient(settings)
```
Comment (jgadling) [inline src/backend/aspen/api/main.py:56]: This is where our fastapi app starts up. We're setting up split pretty early.

File: src/backend/aspen/api/views/users.py [hunk 5]
```diff
@@ -1,19 +1,16 @@
...
 from aspen.api.deps import get_db
-from aspen.api.schemas.users import User as userschema
-from aspen.database.models.usergroup import User
+from aspen.api.schemas.users import UserMeResponse
 
 router = APIRouter()
 
 
-@router.get("/me", response_model=userschema)
+@router.get("/me", response_model=UserMeResponse)
```
Comment (jgadling) [inline src/backend/aspen/api/views/users.py:12]: We're trying to keep the places where we expose split ID's to a minimum -- for now it's only on the /users/me endpoint.

File: src/backend/aspen/database/models/usergroup.py [hunk 6]
```diff
@@ -59,6 +66,7 @@ class User(idbase, DictMixin):  # type: ignore
     # Date of policies (any of Privacy Policy, Terms of Service, etc, etc) the user
     # has last acknowledged. Used to display notification to user when policies change.
     acknowledged_policy_version = Column(Date, nullable=True, default=None)
+    split_id = Column(String, nullable=False, default=generate_split_id)
```
Comment (jgadling) [inline src/backend/aspen/database/models/usergroup.py:69]: Sets the value of a new users's split_id to a random string.

File: src/backend/aspen/util/split.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+from splitio import get_factory
+from splitio.client.client import Client as SplitioClient
+from splitio.exceptions import TimeoutException
+
+
+class SplitClient:
```
Comment (jgadling) [inline src/backend/aspen/util/split.py:6]: Simple wrapper for handling the Split SDK

**Review by lvreynoso — APPROVED** [review]

**Pull request merged**